scene/gui/text_edit.cpp

```diff
@@ -2420,26 +2420,44 @@ void TextEdit::_gui_input(const Ref<InputEvent> &p_gui_input) {
 
 						//simple unindent
 						int cc = cursor.column;
+
+						const int len = text[cursor.line].length();
+						const String &line = text[cursor.line];
+
+						int left;
+						while (left < len && (line[left] == '\t' || line[left] == ' '))
+							left++;
+						cc = MIN(cc, left);
+
+						while (cc < indent_size && cc < left && line[cc] == ' ')
+							cc++;
+
 						if (cc > 0 && cc <= text[cursor.line].length()) {
-							if (text[cursor.line][cursor.column - 1] == '\t') {
-								backspace_at_cursor();
+							if (text[cursor.line][cc - 1] == '\t') {
+								_remove_text(cursor.line, cc - 1, cursor.line, cc);
+								if (cursor.column >= left)
+									cursor_set_column(MAX(0, cursor.column - 1));
+								update();
 							} else {
-								if (cursor.column - indent_size >= 0) {
+								int n = 0;
 
-									bool unindent = true;
-									for (int i = 1; i <= indent_size; i++) {
-										if (text[cursor.line][cursor.column - i] != ' ') {
-											unindent = false;
-											break;
-										}
+								for (int i = 1; i <= MIN(cc, indent_size); i++) {
+									if (line[cc - i] != ' ') {
+										break;
 									}
+									n++;
+								}
 
-									if (unindent) {
-										_remove_text(cursor.line, cursor.column - indent_size, cursor.line, cursor.column);
-										cursor_set_column(cursor.column - indent_size);
-									}
+								if (n > 0) {
+									_remove_text(cursor.line, cc - n, cursor.line, cc);
+									if (cursor.column > left - n)
+										cursor_set_column(MAX(0, left - n));
```

Comment:
Can this not be `cursor.column - n` to keep caret position when unindenting?

Reply:
No, but the code is confusing - I tried to work on the existing code, which always deletes to the left of the cursor and things would be simpler if we'd delete to the right, but that would have changed about everything.

Here are the two cases that can happen:

Case 1:

Take the following text (X is space, T is text, ^ is the cursor position, and `left` is the number of `X`):

```
XXXXXXXXXXTTTTTTT
    ^
```

An unindent happens; `_remove_text` will remove the area to the left of the cursor marked by R:

```
RRRRXXXXXXTTTTTTT
    ^
```

And after removing:

```
XXXXXXTTTTTTT
    ^
```

In this case we do not want to move the cursor, as it is in the white space area and should stay at the same position as before.

Case 2:

Take this text:

```
XXXXXXXXXXTTTTTTT
        ^
```

Removing:

```
XXXXRRRRXXTTTTTTT
        ^
```

```
XXXXXXTTTTTTT
      * ^
```

Now the cursor is in the text area (which begins at `left - n` at `*`) and we need to move the cursor left.

Reply:
Should've been more clear, I meant the `cursor_set_column(MAX(0, left - n));` to `cursor_set_column(MAX(0, cursor.column - n));` else the caret is always going to be moved the start of the text. :)

Reply:
Ah ok, that makes sense, changed.